Sources/Algorithms/Consumers/Consumer.swift

```diff
@@ -0,0 +1,24 @@
+public protocol CollectionConsumer {
+  associatedtype Consumed: Collection
+  func consume(_ consumed: Consumed, subrange: Range<Consumed.Index>) -> Consumed.Index?
```

Comment:
The imperative verb `consume` gives a mutating vibe to it according to the API guidelines.  Are there other naming options that could better align with the semantics of the method?